base/array.jl

```diff
@@ -824,22 +824,23 @@ function collect_to_with_first!(dest, v1, itr, st)
 end
 
 function setindex_widen_up_to(dest::AbstractArray{T}, el, i) where T
-    @inline
     new = similar(dest, promote_typejoin(T, typeof(el)))
     f = first(LinearIndices(dest))
     copyto!(new, first(LinearIndices(new)), dest, f, i-f)
-    @inbounds new[i] = el
+    new[i] = el
     return new
 end
 
 function collect_to!(dest::AbstractArray{T}, itr, offs, st) where T
     # collect to dest array, checking the type of each result. if a result does not
     # match, widen the result type and re-dispatch.
     i = offs
+    len = length(dest)
     while true
         y = iterate(itr, st)
         y === nothing && break
         el, st = y
+        @assert len == length(dest)
         if el isa T
             @inbounds dest[i] = el
```

Comment:
@aviatesk could we make this `@inbounds` conditional on `iterate(itr, st)` being pure and inlined (e.g. a simple load or computation)? That is the only case where this might SIMD, and thus the main case where we might care to eliminate the check here.

(We would like to check also if the user miscomputed `length(itr)` also and so this ran too many or too few iterations, but that seems harder)